system_metrics_collector/test/system_metrics_collector/test_linux_memory_measurement.cpp

```diff
@@ -143,37 +227,210 @@ class LinuxMemoryMeasurementTestFixture : public ::testing::Test
+  {
+    return times_received;
+  }
+
+private:
+  void MetricsMessageCallback(const MetricsMessage & msg)
+  {
+    ASSERT_GT(expected_stats.size(), times_received);
+
+    // check source names
+    EXPECT_EQ(TEST_NODE_NAME, msg.measurement_source_name);
+    EXPECT_EQ("memory_usage", msg.metrics_source);
+
+    // check measurement window
+    // std::chrono::seconds window_sec(msg.window_stop.sec - msg.window_start.sec);
+    // std::chrono::nanoseconds window_nanosec(msg.window_stop.nanosec - msg.window_start.nanosec);
+    // std::chrono::milliseconds window = std::chrono::duration_cast<std::chrono::milliseconds>(
+    //   window_sec + window_nanosec);
+    // EXPECT_GT(5, std::abs(window.count() - PUBLISH_PERIOD.count()));
+
+    // check measurements
+    for (int i = 0; i < STATISTICS_DATA_TYPES.size(); ++i) {
+      EXPECT_EQ(STATISTICS_DATA_TYPES[i], msg.statistics[i].data_type);
+      if (std::isnan(expected_stats[times_received][i].data)) {
+        EXPECT_TRUE(std::isnan(msg.statistics[i].data));
+      } else {
+        EXPECT_DOUBLE_EQ(expected_stats[times_received][i].data, msg.statistics[i].data);
+      }
+    }
+
+    ++times_received;
+  }
+
+  rclcpp::Subscription<MetricsMessage>::SharedPtr subscription;
+  std::array<StatisticDataPoint[STATISTICS_DATA_TYPES.size()], 6> expected_stats;
+  int times_received;
+};
 
 TEST(LinuxMemoryMeasurementTest, testReadInvalidFile)
 {
   const auto s = system_metrics_collector::readFileToString("this_will_fail.txt");
   ASSERT_EQ("", s);
 }
 
-TEST(LinuxMemoryMeasurementTest, testProcessLines)
+TEST(LinuxMemoryMeasurementTest, testprocessMemInfoLines)
 {
-  auto d = system_metrics_collector::processMemInfoLines(EMPTY_SAMPLE);
+  auto d = processMemInfoLines(EMPTY_SAMPLE);
   ASSERT_TRUE(std::isnan(d));
 
-  d = system_metrics_collector::processMemInfoLines(GARBAGE_SAMPLE);
+  d = processMemInfoLines(GARBAGE_SAMPLE);
   ASSERT_TRUE(std::isnan(d));
 
-  d = system_metrics_collector::processMemInfoLines(INCOMPLETE_SAMPLE);
+  d = processMemInfoLines(INCOMPLETE_SAMPLE);
   ASSERT_TRUE(std::isnan(d));
 
-  d = system_metrics_collector::processMemInfoLines(COMPLETE_SAMPLE);
+  d = processMemInfoLines(COMPLETE_SAMPLE);
   ASSERT_DOUBLE_EQ(MEMORY_USED_PERCENTAGE, d);
 
-  d = system_metrics_collector::processMemInfoLines(FULL_SAMPLE);
+  d = processMemInfoLines(FULL_SAMPLE);
   ASSERT_DOUBLE_EQ(MEMORY_USED_PERCENTAGE, d);
 }
+
+TEST_F(LinuxMemoryMeasurementTestFixture, testManualMeasurement) {
+  test_measure_linux_memory->setTestString("");
+  double mem_used_percentage = test_measure_linux_memory->periodicMeasurement();
+  ASSERT_TRUE(std::isnan(mem_used_percentage));
+
+  test_measure_linux_memory->setTestString(FULL_SAMPLE);
+  mem_used_percentage = test_measure_linux_memory->periodicMeasurement();
+  ASSERT_DOUBLE_EQ(MEMORY_USED_PERCENTAGE, mem_used_percentage);
+}
+
+TEST_F(LinuxMemoryMeasurementTestFixture, testPublishMetricsMessage)
+{
+  ASSERT_NE(test_measure_linux_memory, nullptr);
```

Comment:
use the assert for nullptr?

Reply:
It turned out there isn't an assert specifically for `nullptr`.